pkg/zng/docs/spec.md

```diff
@@ -1,96 +1,268 @@
+## 1. The ZNG data model
 
-Directives indicate how subsequent values in the ZSON stream are interpreted.
-A value is a regular value if the most recent preceding directive in the stream
-is a regular directive; otherwise, it is a legacy value.
+ZNG encodes a sequence of one or more typed data values comprising a stream.
+The stream of values is interleaved with control messages
+that provide type definitions and other metadata.  The type of
+a particular data value is specified by its "type code", which is an
+integer identifier representing either a built-in type or
+a dynamic type definition that occurred previously in the stream.
 
-Any line beginning with `#` that does not conform with the syntax of a
-directive is an error.
-When errors are encountered parsing ZSON, an implementation should return a
-corresponding error and allow ZSON parsing to proceed if desired.
+The ZNG type system comprises the standard set of scalar types like integers,
+floating point, strings, byte arrays, etc. as well as composite types
+like records, arrays, and sets arranged from other types.
+
+For example, a ZNG stream representing the single string "hello world"
+looks like this:
+```
+9:hello, world
+```
+Here, the type code is the integer "9" representing the string type and the data
+value "hello, world" is an instance of string.
+
+ZNG gets more interesting when data types are interleaved in the stream.
+For example,
+```
+9:hello, world
+4:42
+9:there's a fly in my soup!
+9:no, there isn't.
+4:3
+```
+where type code 4 represents an integer.  This encoding represents the string of
+values:
+```
+"hello, world"
+42
+"there's a fly in my soup!"
+"no, there isn't."
+3
+```
+ZNG streams are often comprised as a sequence of records, which works well to provide
+an efficient representation of structured logs.  In this case, a new type code is
+needed to define the schema for each distinct record.  To define a new
+type, the "#" syntax is used.  For example,
+logs from the open-source Zeek system might look like this
+```
+#23:ip=addr
+#24:record[_path:string,ts:time,uid:string,id:record[orig_h:addr,orig_p:port,resp_h:addr,resp_p:port]...
+#25:record[_path:string,ts:time,uid:string,id:record[orig_h:addr,orig_p:port,resp_h:addr,resp_p:port]...
+24:conn;1425565514.419939;CogZFI3py5JsFZGik;[192.168.1.1:;80/tcp;192.168.1.2;8080;]...
+25:dns;1425565514.419987;CogZFI3py5JsFZGik;[192.168.1.1:;5353/udp;192.168.1.2;5353;]...
+```
+Note that the value encoding need not refer to the field names and types as that is
+completely captured by the type code.  Values merely encode the value
+information consistent with the referenced type code.
+
+## 2. ZNG Binary Format (BZNG)
+
+The BZNG binary format is based on machine-readable data types with an
+encoding methodology inspired by
+[Protocol Buffers](https://developers.google.com/protocol-buffers).
+
+A BZNG stream comprises a sequence control messages and value messages
```

Comment:
s/a sequence control messages/a sequence of control messages/

Reply:
+1